compiler/next/lib/frontend/Parser/ParserContextImpl.h

```diff
@@ -226,6 +226,27 @@ CommentsAndStmt ParserContext::finishStmt(Expression* e) {
   return makeCommentsAndStmt(NULL, e);
 }
 
+ParserExprList*
+ParserContext::blockToParserExprList(YYLTYPE lbrLoc, YYLTYPE rbrLoc,
+                                     ParserExprList* body) {
+  /*
+  // create a list of the statements
+  ParserExprList* lst = $2;
+  assert(lst != nullptr);
+  // add also any comments appearing after all statements
+  // and before the closing close bracket.
+  context->appendList(lst, context->gatherComments(@3));
+  */
+
+  ParserExprList* ret = body != nullptr ? body : new ParserExprList();
+
+  // Add any comments appearing after all statements and before the closing
+  // close bracket.
+  this->appendList(ret, this->gatherComments(rbrLoc));
```

Comment:
What about comments occuring before the `{` ?

We had this
```
    // comments from before the opening bracket will have been
    // gathered into stmt_ls when that was parsed, so pull out any
    // comments that occur before the opening bracket.
    // (This would be simpler to handle with midrule actions but those
    //  lead to parser conflicts).
    CommentsAndStmt cs = {0};
    cs.comments = context->gatherCommentsFromList(lst, @1);
```

I'd expect this function to return a CommentsAndList, where the comments from before the `{` are stored in the `.comments` of it, and comments inside the `{ }` are stored (along with the non-comment statements) in the `.list` field. (CommentsAndList would be a new type but you could add a bunch of overloads in ParserContext to make it as easy to work with).


Oh, I see. You have

```
   // Pull up any comments that appear before the TLOCAL keyword.
    cs.comments = context->gatherCommentsFromList($3.exprList, @1);
```

in the parser action.

I think the new type CommentsAndList would be nice but it's not critical. You could fold the `usesDo` into the same type, though.

Reply:
The reason why I don't bother doing that is because then I have to lift the comments up again in the parser action for local, which has to use the location of `TLOCAL` to do so. In https://github.com/chapel-lang/chapel/pull/17647/commits/9bf9369a36a4b21c825a3b14a67e84abb262009f I went ahead and streamlined things to about as pretty as they're going to be.

Reply:
OK, good point.